pkg/csr/csr.go

```diff
@@ -0,0 +1,357 @@
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/openshift/windows-machine-config-operator/pkg/instances"
+	"github.com/openshift/windows-machine-config-operator/pkg/parser"
+	"github.com/openshift/windows-machine-config-operator/pkg/secrets"
+	"github.com/openshift/windows-machine-config-operator/pkg/signer"
+	"github.com/openshift/windows-machine-config-operator/pkg/windows"
+)
+
+//+kubebuilder:rbac:groups="certificates.k8s.io",resources=certificatesigningrequests,verbs=update/approval
+//+kubebuilder:rbac:groups="certificates.k8s.io",resources=certificatesigningrequests,verbs=get;list;watch
+//+kubebuilder:rbac:groups="certificates.k8s.io",resources=signers,verbs=Approve,resourceNames=kubernetes.io/kube-apiserver-client-kubelet;kubernetes.io/kubelet-serving
+
+// kubeletClientUsages contains the permitted key usages from a kube-apiserver-client-kubelet signer
+var kubeletClientUsages = []certificates.KeyUsage{
+	certificates.UsageKeyEncipherment,
+	certificates.UsageDigitalSignature,
+	certificates.UsageClientAuth,
+}
+
+// kubeletServerUsages contains the permitted key usages from a kubelet-serving signer
+var kubeletServerUsages = []certificates.KeyUsage{
+	certificates.UsageKeyEncipherment,
+	certificates.UsageDigitalSignature,
+	certificates.UsageServerAuth,
+}
+
+const (
+	nodeGroup          = "system:nodes"
+	nodeUserName       = "system:node"
+	NodeUserNamePrefix = nodeUserName + ":"
+	systemPrefix       = "system:authenticated"
+)
+
+// Approver holds the information required to approve a node CSR
+type Approver struct {
+	// client is the cache client
+	client client.Client
+	// k8sclientset holds the kube client needed for updating CSR approval conditions
+	k8sclientset *kubernetes.Clientset
+	// csr holds the pointer to the CSR request
+	csr      *certificates.CertificateSigningRequest
+	log      logr.Logger
+	recorder record.EventRecorder
+	// namespace is the namespace in which CSR's are present
+	namespace string
+}
+
+// NewApprover returns a pointer to the Approver
+func NewApprover(client client.Client, clientSet *kubernetes.Clientset, csr *certificates.CertificateSigningRequest,
+	log logr.Logger, recorder record.EventRecorder, watchNamespace string) (*Approver, error) {
+	if client == nil || csr == nil || clientSet == nil {
+		return nil, errors.New("kubernetes client, clientSet or CSR should not be nil")
+	}
+	return &Approver{client,
+		clientSet,
+		csr,
+		log,
+		recorder,
+		watchNamespace}, nil
+}
+
+// Approve approves a CSR by updating it's status conditions to true if it is a valid CSR
+func (a *Approver) Approve() error {
+	if a.k8sclientset == nil {
+		return errors.New("Kubernetes clientSet should not be nil")
+	}
+
+	if validated, err := a.validateCSRContents(); !validated {
+		a.log.Info("CSR contents are invalid for approval by WMCO CSR Approver", "CSR", a.csr.Name)
+		return errors.Wrapf(err, "could not validate CSR %s contents for approval", a.csr.Name)
+	}
```

Comment:
This doesnt seem to be filtering things out correctly. If err == nil and !validated it should return nil here? If not its going to cause a requeue on something we dont want to approve